src/Hamiltonians/correlation_functions.jl

```diff
@@ -236,3 +236,153 @@ function get_offdiagonal(
+- [G2 Correlator Example](https://joachimbrand.github.io/Rimu.jl/previews/PR227/generated/G2-example.html)
+
+
+# See also 
+
+* [`HubbardReal1D`](@ref)
+* [`G2RealCorrelator`](@ref)
+* [`AbstractHamiltonian`](@ref)
+* [`AllOverlaps`](@ref)
+
+"""
+struct SuperfluidCorrelator{D} <: AbstractHamiltonian{Float64}
+end
+
+SuperfluidCorrelator(d::Int) = SuperfluidCorrelator{d}()
+
+function Base.show(io::IO, ::SuperfluidCorrelator{D}) where {D}
+    print(io, "SuperfluidCorrelator($D)")
+end
+
+function num_offdiagonals(::SuperfluidCorrelator, add::SingleComponentFockAddress)
+    return num_occupied_modes(add)
+end
+
+function get_offdiagonal(::SuperfluidCorrelator{D}, add::SingleComponentFockAddress, chosen) where {D}
+    src = find_occupied_mode(add, chosen)
+    dst = find_mode(add, mod1(src.mode + D, num_modes(add)))
+    address, value = excitation(add, (dst,), (src,))
+    return address, value / num_modes(add)
+end
+
+function diagonal_element(::SuperfluidCorrelator{0}, add::SingleComponentFockAddress)
+    return num_particles(add) / num_modes(add)
+end
+function diagonal_element(::SuperfluidCorrelator{D}, add::SingleComponentFockAddress) where {D}
+    return 0.0
+end
+
+
+"""
+    StringCorrelator(d::Int) <: AbstractHamiltonian{Float64}
+
+Two-body operator for string correlation between sites separated by `d` with `0 ≤ d < M`
+
+```math
+    \\hat{C}_{\\text{string}}(d) = \\frac{1}{M} \\sum_{j}^{M} \\delta n_j (e^{i \\pi \\sum_{j \\leq k < j + d} \\delta n_k}) \\delta n_{j+d}
+```
+Here ``\\bar{n} = N/M`` is the mean filling number of lattice sites with ``N`` and ``M`` being the number of particles and number of lattice sites respectively. ``\\delta \\hat{n}_j = \\hat{n}_j - \\bar{n}`` is the boson number deviation from the mean filling number.
+
+Assumes a one-dimensional lattice with periodic boundary conditions. 
+
+# Arguments
+- `d::Integer`: distance between sites. 
+
+# See also 
+
+* [`HubbardReal1D`](@ref)
+* [`G2RealCorrelator`](@ref)
+* [`SuperfluidCorrelator`](@ref)
+* [`AbstractHamiltonian`](@ref)
+* [`AllOverlaps`](@ref)
+
+"""
+struct StringCorrelator{D} <: AbstractHamiltonian{Float64}
+end
+
+StringCorrelator(d::Int) = StringCorrelator{d}()
+
+function Base.show(io::IO, ::StringCorrelator{D}) where {D}
+    print(io, "StringCorrelator($D)")
+end
+
+LOStructure(::Type{<:StringCorrelator}) = IsDiagonal()
+
+function diagonal_element(::StringCorrelator{0}, add::SingleComponentFockAddress)
+    M = num_modes(add)
+    N = num_particles(add)
+    n̄ = N/M
+    v = onr(add)
+
+    result = 0.0
+    for i in eachindex(v)
+        result += (v[i] - n̄)^2
+    end
+
+    return result / M
+end
+
+num_offdiagonals(::StringCorrelator, ::SingleComponentFockAddress) = 0
+
+function diagonal_element(::StringCorrelator{D}, add::SingleComponentFockAddress) where {D}
+    M = num_modes(add)
+    N = num_particles(add)
+    d = mod(D, M)
+
+    if iszero(N % M)
```

Comment:
```suggestion
    if !ismissing(N) && iszero(N % M)
```
With `OccupationNumberFS` the value of `N` could be `missing`.